### DISCUSSION TITLE
[Suggestion] Add index.* to every library directory

### DISCUSSION THREAD

**Discussion by l-spiecker** [papis/papis#1170] [Ideas]
The current concept of having an info.yml is nice as it contains very much information in a dense file, but it is not as readable as a markup file especially if you use a file explorer or a scm like github which renders markup files nicely. It would also allow building a papis library as static html with a documentation system like sphinx.

Papis could create and update an index.md/rst/html automatically on every yml change for every document directory upwards to the root of the library just by iterating over folders and running a templating language like jinja2. It could be also done from a papis plugin or an small external script but first level integrations are often better.

Maybe there is already such an option or tool for it, then let me know.

**Comment by alexfikl** [comment]
I think `papis-html` might do what you want?
https://github.com/papis/scripts/tree/master/papis-html

Haven't tried it myself and don't know if it still works, but it should be a good starting point!

**Comment by l-spiecker** [comment]
Not bad, but i just wanted to have an index.md in every folder. I made me now a simple script, which runs every jinja template with every info.yml and also with a list for the root directory. I didn't made it yet as a papis script as i wanted to place it next to the bib folder and not under my home.

Maybe it helps someone else here.

It needs _templates, of course:
```
_templates/document/[index.md.jinja, ...]
_templates/directory/[index.md.jinja, ...]
```

```
from pathlib import Path

import yaml
from jinja2 import Environment, FileSystemLoader

path = Path(__file__).parent / "../../papers"

environment = Environment(loader=FileSystemLoader("_templates/"))

dirs = {}
for info_path in Path(path).glob("**/info.yaml"):
    with open(info_path) as file:
        info = yaml.load(file, yaml.SafeLoader)
    info['year'] = int(info['year'])  # Correction for jinja sorting
    dirs[info_path.parent.parent] = dirs.get(info_path.parent.parent, []) + [
        {'path': info_path.parent.name, 'info': info}]
    for template_file in environment.list_templates(filter_func=lambda s: Path(s).parts[0] == 'document'):
        template = environment.get_template(template_file)
        with open(info_path.parent / Path(template.name).stem, mode="w", encoding="utf-8") as file:
            file.write(template.render(**info))

for dir_path, documents in dirs.items():
    for template_file in environment.list_templates(filter_func=lambda s: Path(s).parts[0] == 'directory'):
        template = environment.get_template(template_file)
        with open(dir_path / Path(template.name).stem, mode="w", encoding="utf-8") as file:
            file.write(template.render(documents=documents, name=dir_path.name))
```

**Comment by alejandrogallo** [comment]
That is a cool idea and it shows future users how "easy" it is to code up some /ad hoc/ solution for your use case. Thank you for sharing and feel free to share some end product of your experimentation!
